Hello Deb!

I’m happy to review your project today.
I will mark your mistakes and give you some hints how it is possible to fix them. We are getting ready for real job, where your team leader/senior colleague will do exactly the same. Don't worry and study with pleasure! 

Below you will find my comments - **please do not move, modify or delete them**.

You can find my comments in green, yellow or red boxes like this:

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Success. Everything is done succesfully.
</div>

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Remarks. Some recommendations.
</div>

<div class="alert alert-block alert-danger">

<b>Reviewer's comment</b> <a class="tocSkip"></a>

Needs fixing. The block requires some corrections. Work can't be accepted with the red comments.
</div>

You can answer me by using this:

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>

Text here.
</div>

## Basic Python - Project <a id='intro'></a>

## Introduction <a id='intro'></a>
In this project, you will work with data from the entertainment industry. You will study a dataset with records on movies and shows. The research will focus on the "Golden Age" of television, which began in 1999 with the release of *The Sopranos* and is still ongoing.

The aim of this project is to investigate how the number of votes a title receives impacts its ratings. The assumption is that highly-rated shows (we will focus on TV shows, ignoring movies) released during the "Golden Age" of television also have the most votes.

### Stages 
Data on movies and shows is stored in the `/datasets/movies_and_shows.csv` file. There is no information about the quality of the data, so you will need to explore it before doing the analysis.

First, you'll evaluate the quality of the data and see whether its issues are significant. Then, during data preprocessing, you will try to account for the most critical problems.
 
Your project will consist of three stages:
 1. Data overview
 2. Data preprocessing
 3. Data analysis

## Stage 1. Data overview <a id='data_review'></a>

Open and explore the data.

You'll need `pandas`, so import it.

In [62]:
# importing pandas
import pandas as pd

Read the `movies_and_shows.csv` file from the `datasets` folder and save it in the `df` variable:

In [63]:
# reading the files and storing them to df
df = pd.read_csv('/datasets/movies_and_shows.csv')
#print(df['  Type'].sort_values(ascending=True).unique())

Print the first 10 table rows:

In [64]:
# obtaining the first 10 rows from the df table
# hint: you can use head() and tail() in Jupyter Notebook without wrapping them into print()
display(df.head(10))

name                      Character   r0le        TITLE   Type  \
0   Robert De Niro                  Travis Bickle  ACTOR  Taxi Driver  MOVIE   
1     Jodie Foster                  Iris Steensma  ACTOR  Taxi Driver  MOVIE   
2    Albert Brooks                            Tom  ACTOR  Taxi Driver  MOVIE   
3    Harvey Keitel        Matthew 'Sport' Higgins  ACTOR  Taxi Driver  MOVIE   
4  Cybill Shepherd                          Betsy  ACTOR  Taxi Driver  MOVIE   
5      Peter Boyle                         Wizard  ACTOR  Taxi Driver  MOVIE   
6   Leonard Harris      Senator Charles Palantine  ACTOR  Taxi Driver  MOVIE   
7   Diahnne Abbott                Concession Girl  ACTOR  Taxi Driver  MOVIE   
8      Gino Ardito             Policeman at Rally  ACTOR  Taxi Driver  MOVIE   
9  Martin Scorsese  Passenger Watching Silhouette  ACTOR  Taxi Driver  MOVIE   

   release Year              genres  imdb sc0re  imdb v0tes  
0          1976  ['drama', 'crime']         8.2    808582.0  
1          1976  ['drama', 'crime']         8.2    808582.0  
2          1976  ['drama', 'crime']         8.2    808582.0  
3          1976  ['drama', 'crime']         8.2    808582.0  
4          1976  ['drama', 'crime']         8.2    808582.0  
5          1976  ['drama', 'crime']         8.2    808582.0  
6          1976  ['drama', 'crime']         8.2    808582.0  
7          1976  ['drama', 'crime']         8.2    808582.0  
8          1976  ['drama', 'crime']         8.2    808582.0  
9          1976  ['drama', 'crime']         8.2    808582.0

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

We do not need print() in Jupyter notebook.
</div>

Obtain the general information about the table with one command:

In [65]:
# obtaining general information about the data in df
df.info()
#df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85579 entries, 0 to 85578
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0      name       85579 non-null  object 
 1   Character     85579 non-null  object 
 2   r0le          85579 non-null  object 
 3   TITLE         85578 non-null  object 
 4     Type        85579 non-null  object 
 5   release Year  85579 non-null  int64  
 6   genres        85579 non-null  object 
 7   imdb sc0re    80970 non-null  float64
 8   imdb v0tes    80853 non-null  float64
dtypes: float64(2), int64(1), object(6)
memory usage: 5.9+ MB


<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

We need to use .info() here.
</div>

The table contains nine columns. The majority store the same data type: object. The only exceptions are `'release Year'` (int64 type), `'imdb sc0re'` (float64 type) and `'imdb v0tes'` (float64 type). Scores and votes will be used in our analysis, so it's important to verify that they are present in the dataframe in the appropriate numeric format. Three columns (`'TITLE'`, `'imdb sc0re'` and `'imdb v0tes'`) have missing values.

According to the documentation:
- `'name'` — actor/director's name and last name
- `'Character'` — character played (for actors)
- `'r0le '` — the person's contribution to the title (it can be in the capacity of either actor or director)
- `'TITLE '` — title of the movie (show)
- `'  Type'` — show or movie
- `'release Year'` — year when movie (show) was released
- `'genres'` — list of genres under which the movie (show) falls
- `'imdb sc0re'` — score on IMDb
- `'imdb v0tes'` — votes on IMDb

We can see three issues with the column names:
1. Some names are uppercase, while others are lowercase.
2. There are names containing whitespace.
3. A few column names have digit '0' instead of letter 'o'. 


### Conclusions <a id='data_review_conclusions'></a> 

Each row in the table stores data about a movie or show. The columns can be divided into two categories: the first is about the roles held by different people who worked on the movie or show (role, name of the actor or director, and character if the row is about an actor); the second category is information about the movie or show itself (title, release year, genre, imdb figures).

It's clear that there is sufficient data to do the analysis and evaluate our assumption. However, to move forward, we need to preprocess the data.

## Stage 2. Data preprocessing <a id='data_preprocessing'></a>
Correct the formatting in the column headers and deal with the missing values. Then, check whether there are duplicates in the data.

In [66]:
# the list of column names in the df table
df.columns

Index(['   name', 'Character', 'r0le', 'TITLE', '  Type', 'release Year',
       'genres', 'imdb sc0re', 'imdb v0tes'],
      dtype='object')

Change the column names according to the rules of good style:
* If the name has several words, use snake_case
* All characters must be lowercase
* Remove whitespace
* Replace zero with letter 'o'

In [67]:
# renaming columns
df.columns = df.columns.str.lower()
df.columns = df.columns.str.strip()
df.columns = df.columns.str.replace(' ', '_')
df = df.rename(columns={
    'r0le':'role',
    'imdb_sc0re': 'imdb_score',
    'imdb_v0tes': 'imdb_votes'
    
    })
   
#df.columns

Check the result. Print the names of the columns once more:

In [68]:
# checking result: the list of column names
df.columns
#print(df.type)

Index(['name', 'character', 'role', 'title', 'type', 'release_year', 'genres',
       'imdb_score', 'imdb_votes'],
      dtype='object')

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Well done!
</div>

### Missing values <a id='missing_values'></a>
First, find the number of missing values in the table. To do so, combine two `pandas` methods:

In [69]:
# calculating missing values
df.isna().sum()

name               0
character          0
role               0
title              1
type               0
release_year       0
genres             0
imdb_score      4609
imdb_votes      4726
dtype: int64

Not all missing values affect the research: the single missing value in `'title'` is not critical. The missing values in columns `'imdb_score'` and `'imdb_votes'` represent around 6% of all records (4,609 and 4,726, respectively, of the total 85,579). This could potentially affect our research. To avoid this issue, we will drop rows with missing values in the `'imdb_score'` and `'imdb_votes'` columns.

In [70]:
# dropping rows where columns with title, scores and votes have missing values
df = df.dropna(subset=['imdb_score', 'imdb_votes', 'title'])

Make sure the table doesn't contain any more missing values. Count the missing values again.

In [71]:
# counting missing values
df.isna().sum()

name            0
character       0
role            0
title           0
type            0
release_year    0
genres          0
imdb_score      0
imdb_votes      0
dtype: int64

### Duplicates <a id='duplicates'></a>
Find the number of duplicate rows in the table using one command:

In [72]:
# counting duplicate rows
print(df.duplicated().sum())

6994


Review the duplicate rows to determine if removing them would distort our dataset.

In [73]:
# Produce table with duplicates (with original rows included) and review last 5 rows
# All occurances of a duplicated value are marked as True, because of "keep=False"
dupe_rows = df[df.duplicated()]
display(dupe_rows.tail(55))
dupe_rows.shape

name                     character      role  \
85515          Fulu Mugovhani                       Lindiwe     ACTOR   
85516            Simo Mogwaza                          Beki     ACTOR   
85517          Thabang Moleya                       unknown  DIRECTOR   
85518             Ai Leng Ong                 Jack's Mother     ACTOR   
85519            Chew Kin-Wah                 Jack's Father     ACTOR   
85520         Harith Iskander             Sharifah's Father     ACTOR   
85521             Erra Fazira              Sharifa's Mother     ACTOR   
85522              Teddy Chan                       unknown  DIRECTOR   
85524          Maymay Entrata                       unknown     ACTOR   
85525           Edward Barber                       unknown     ACTOR   
85526            Snooky Serna                       unknown     ACTOR   
85528            Pepe Herrera                       unknown     ACTOR   
85529           Chie Filomeno                       unknown     ACTOR   
85530         Alora Mae Sasam                       unknown     ACTOR   
85531         Iggy Boy Flores                       unknown     ACTOR   
85532             Neil Coleta                       unknown     ACTOR   
85533              Cj Salonga                       unknown     ACTOR   
85534             Gold Azeron                       unknown     ACTOR   
85535        Christine Samson                       unknown     ACTOR   
85536          Takuhei Kaneko                       unknown     ACTOR   
85537          Big Mac Andaya                       unknown     ACTOR   
85538          Miko Pe?ñaloza                       unknown     ACTOR   
85539          Barry Gonzalez                       unknown  DIRECTOR   
85540             Ahmed Hatem                       unknown     ACTOR   
85541             Jamila Awad                       unknown     ACTOR   
85542      Mahmoud Al-Bezzawy                       unknown     ACTOR   
85543                 Sabrien                       unknown     ACTOR   
85544          Zeinab Ghareeb                       unknown     ACTOR   
85545           Marwan Younis                       unknown     ACTOR   
85546              Nada Akram                       unknown     ACTOR   
85547             Asmaa Galal                       unknown     ACTOR   
85548        Hind Abdulhaleem                       unknown     ACTOR   
85549     Passant El Nabarawy                       unknown     ACTOR   
85550          Sherif Desouqy                       unknown     ACTOR   
85551             Seif Zahran                       unknown     ACTOR   
85552           Mohamed Moula                       unknown     ACTOR   
85553             Ahmed Bahaa                       unknown     ACTOR   
85554        Mohamed El-Arkan                       unknown     ACTOR   
85555           Mohamed Bakir                       unknown  DIRECTOR   
85556     Richard Mofe-Damijo                    Mr. George     ACTOR   
85557               Ego Nwosu                         Kaima     ACTOR   
85558  Keppy Ekpenyong-Bassey                          Akin     ACTOR   
85559        Ademola Adedoyin                         Tunji     ACTOR   
85561            Segun Arinze                       unknown     ACTOR   
85562           Nse Ikpe-Etim                           Ame     ACTOR   
85563                Tina Mba                         Mummy     ACTOR   
85565           Bofie Itombra                         Kamsi     ACTOR   
85566             Kameel Audu                        Andrew     ACTOR   
85567           Seyi Babatope                       unknown  DIRECTOR   
85568        Yedinson Fl??rez  Yimmy Bar??n / Paola Linares     ACTOR   
85569          Jessica Cediel               Liliana Navarro     ACTOR   
85570     Javier Gardeaz?­bal      Agust??n "Peluca" Ort??z     ACTOR   
85571           Carla Giraldo                Valery Reinoso     ACTOR   
85572     Ana Mar??a S?­nchez                       Lourdes     ACTOR   
85577     

(6994, 9)

There are two clear duplicates in the printed rows. We can safely remove them.
Call the `pandas` method for getting rid of duplicate rows:

In [74]:
# removing duplicate rows
# df = dupe_rows.drop_duplicates()
# Remove duplicates based on all columns, excluding the 'type' column
#columns_to_check = df.columns.difference(['type'])
#df = df.drop_duplicates(subset=columns_to_check)

df.drop_duplicates(inplace=True)

print("SHAPE:")
print(df.shape)

#print("Unique types after dropping duplicates:", sorted(df['type'].unique()))

SHAPE:
(73859, 9)


Check for duplicate rows once more to make sure you have removed all of them:

In [75]:
# checking for duplicates
print(df.duplicated().sum())

0


<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Good job.
</div>

Now get rid of implicit duplicates in the `'type'` column. For example, the string `'SHOW'` can be written in different ways. These kinds of errors will also affect the result.

Print a list of unique `'type'` names, sorted in alphabetical order. To do so:
* Retrieve the intended dataframe column 
* Apply a sorting method to it
* For the sorted column, call the method that will return all unique column values

In [76]:
# viewing unique type names. BEFORE.
print(df['type'].sort_values(ascending=True).unique())

['MOVIE' 'SHOW' 'movies' 'shows' 'the movie' 'tv' 'tv series' 'tv show'
 'tv shows']


Look through the list to find implicit duplicates of `'show'` (`'movie'` duplicates will be ignored since the assumption is about shows). These could be names written incorrectly or alternative names of the same genre.

You will see the following implicit duplicates:
* `'shows'`
* `'SHOW'`
* `'tv show'`
* `'tv shows'`
* `'tv series'`
* `'tv'`

To get rid of them, declare the function `replace_wrong_show()` with two parameters: 
* `wrong_shows_list=` — the list of duplicates
* `correct_show=` — the string with the correct value

The function should correct the names in the `'type'` column from the `df` table (i.e., replace each value from the `wrong_shows_list` list with the value in `correct_show`).

In [77]:
# function for replacing implicit duplicates
def replace_wrong_show(wrong_shows_list, correct_show):
    for w in wrong_shows_list:
        df['type'] = df['type'].replace(w, correct_show)
        

Call `replace_wrong_show()` and pass it arguments so that it clears implicit duplicates and replaces them with `SHOW`:

In [78]:
# removing implicit duplicates
# List of incorrect show names
wrong_shows_list = ["shows", "tv show", "tv shows", "tv series", "tv"]
correct_show = "SHOW"
# call function
replace_wrong_show(wrong_shows_list, correct_show)

Make sure the duplicate names are removed. Print the list of unique values from the `'type'` column:

In [79]:
# viewing unique genre names
print(df['type'].unique())

['MOVIE' 'the movie' 'SHOW' 'movies']


<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Correct.
</div>

### Conclusions <a id='data_preprocessing_conclusions'></a>
We detected three issues with the data:

- Incorrect header styles
- Missing values
- Duplicate rows and implicit duplicates

The headers have been cleaned up to make processing the table simpler.

All rows with missing values have been removed. 

The absence of duplicates will make the results more precise and easier to understand.

Now we can move on to our analysis of the prepared data.

## Stage 3. Data analysis <a id='hypotheses'></a>

Based on the previous project stages, you can now define how the assumption will be checked. Calculate the average amount of votes for each score (this data is available in the `imdb_score` and `imdb_votes` columns), and then check how these averages relate to each other. If the averages for shows with the highest scores are bigger than those for shows with lower scores, the assumption appears to be true.

Based on this, complete the following steps:

- Filter the dataframe to only include shows released in 1999 or later.
- Group scores into buckets by rounding the values of the appropriate column (a set of 1-10 integers will help us make the outcome of our calculations more evident without damaging the quality of our research).
- Identify outliers among scores based on their number of votes, and exclude scores with few votes.
- Calculate the average votes for each score and check whether the assumption matches the results.

To filter the dataframe and only include shows released in 1999 or later, you will take two steps. First, keep only titles published in 1999 or later in our dataframe. Then, filter the table to only contain shows (movies will be removed).

In [80]:
df_SHOW_newerThan1999 = df[(df['release_year'] >= 1999) & (df['type'] == 'SHOW')].copy()
print("SHAPE")
print(df_SHOW_newerThan1999.shape)

SHAPE
(13430, 9)


In [81]:
# using conditional indexing modify df so it has only titles released after 1999 (with 1999 included)
# give the slice of dataframe new name
# only include SHOW's from 1999 or later
#df_SHOW_newerThan1999 = df[(df['release_year'] >= 1999) & (df['type'] == 'SHOW')].copy()

# Group scores into buckets, and round. Low , medium, high scores.
#df_SHOW_newerThan1999.loc[:, 'imdb_score_bucket_1-3'] = (df_SHOW_newerThan1999['imdb_score'] <= 3)
#df_SHOW_newerThan1999.loc[:, 'imdb_score_bucket_1-3'] = df_SHOW_newerThan1999['imdb_score_bucket_1-3'].round()
 
#df_SHOW_newerThan1999.loc[:, 'imdb_score_bucket_4-6'] = (df_SHOW_newerThan1999['imdb_score'] > 3) & (df_SHOW_newerThan1999['imdb_score'] <= 6)
#df_SHOW_newerThan1999.loc[:, 'imdb_score_bucket_4-6'] = df_SHOW_newerThan1999['imdb_score_bucket_4-6'].round()
 
#df_SHOW_newerThan1999.loc[:, 'imdb_score_bucket_7-10'] = (df_SHOW_newerThan1999['imdb_score'] > 6) & (df_SHOW_newerThan1999['imdb_score'] <= 10)
#df_SHOW_newerThan1999.loc[:, 'imdb_score_bucket_7-10'] = df_SHOW_newerThan1999['imdb_score_bucket_7-10'].round()

 
# Step 3: Exclude scores with fewer votes (fewer than 500)
#df_SHOW_newerThan1999 = df_SHOW_newerThan1999[df_SHOW_newerThan1999['imdb_votes'] >= 500]
 
def high_med_low(bucket):
    if bucket <= 3:
        return 'low'
    if bucket > 3 and bucket <= 6:
        return 'medium'
    if bucket > 6 and bucket <= 10:
        return 'high'
        
  
df_SHOW_newerThan1999['imdb_rating'] = df_SHOW_newerThan1999['imdb_score'].apply(high_med_low)
    
# Step 4: Calculate the average votes for each score bucket
avg_imdb_score_1_3 = df_SHOW_newerThan1999[df_SHOW_newerThan1999['imdb_rating'] == 'low']['imdb_votes'].mean()
avg_imdb_score_4_6 = df_SHOW_newerThan1999[df_SHOW_newerThan1999['imdb_rating'] == 'medium']['imdb_votes'].mean()
avg_imdb_score_7_10 = df_SHOW_newerThan1999[df_SHOW_newerThan1999['imdb_rating'] == 'high']['imdb_votes'].mean()



print("SHAPE")
print(df_SHOW_newerThan1999.shape)

# Print the result for review, and round to 2 places
# Round the average votes to 2 decimal places
avg_imdb_score_1_3 = round(avg_imdb_score_1_3, 2)
avg_imdb_score_4_6 = round(avg_imdb_score_4_6, 2)
avg_imdb_score_7_10 = round(avg_imdb_score_7_10, 2)

print("Average votes for IMDb score between 1 and 3:", avg_imdb_score_1_3)
print("Average votes for IMDb score between 4 and 6:", avg_imdb_score_4_6)
print("Average votes for IMDb score between 7 and 10:", avg_imdb_score_7_10)
print('\n')
print("Assumption is true... highly-rated shows also have the most votes.")

# Create some "buckets" to group imdb scoring
#imdb_score_bucket_1_3 = 0
#imdb_score_bucket_4_6 = 0
#imdb_score_bucket_7_10 = 0

SHAPE
(13430, 10)
Average votes for IMDb score between 1 and 3: 1801.87
Average votes for IMDb score between 4 and 6: 3070.66
Average votes for IMDb score between 7 and 10: 23940.02


Assumption is true... highly-rated shows also have the most votes.


<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Great additional analysis!
</div>

In [82]:
# repeat conditional indexing so df has only shows (movies are removed as result)
df = df_SHOW_newerThan1999[df_SHOW_newerThan1999['type'] == 'SHOW']
display(df.tail(50))


name       character      role  \
77653         Mawin Taweepol             Win     ACTOR   
77654          Nat Sakdatorn         unknown     ACTOR   
77655       Maneerat Kam-Uan              Ae     ACTOR   
77656      Rudklao Amratisha         unknown     ACTOR   
77657        Jaturong Mokjok         unknown     ACTOR   
77658       Pisamai Wilaisak         unknown     ACTOR   
77659    Sarawut Wichiensarn         unknown  DIRECTOR   
83911        Heather Hemmens         unknown     ACTOR   
84679           Artur Pontek   Kayko (voice)     ACTOR   
84680          Micha? Piela  Kokosh (voice)     ACTOR   
84681         Maciej Kosmala  Milus ( voice)     ACTOR   
84824            Abla Fahita         unknown     ACTOR   
84825                   Caro         unknown     ACTOR   
84826           Bassem Samra           Ameen     ACTOR   
84827            Donia Maher         unknown     ACTOR   
84828          Rehab Hussein         unknown     ACTOR   
84829              Mnaa Adel         unknown     ACTOR   
84830          Bossy Shalaby         unknown     ACTOR   
84831          Zeina Mansour         unknown     ACTOR   
84832         Osama Abdullah         unknown     ACTOR   
84833     Ahmed Saad (Mesho)         unknown     ACTOR   
84834            Fekri Salim         unknown     ACTOR   
84912        M??nica Naranjo     Self - Host     ACTOR   
85068          Robby Collins         unknown     ACTOR   
85314        Shubham Agarwal            Self     ACTOR   
85315        Lokesh Deshmukh            Self     ACTOR   
85316          Mukul Sankule            Self     ACTOR   
85317        Kartikeya Singh            Self     ACTOR   
85318        Adarsh Upadhyay            Self     ACTOR   
85319           Kevin Banker            Self     ACTOR   
85320      Gokul J. Krishnan            Self     ACTOR   
85321      Pradeep S. Gautam            Self     ACTOR   
85322      Biswa Kalyan Rath            Self     ACTOR   
85323           Poonam Gupta            Self     ACTOR   
85324         Spandana Silla            Self     ACTOR   
85325           Pratik Patra         unknown  DIRECTOR   
85326           Prashant Raj         unknown  DIRECTOR   
85398          Javed Jaffrey         unknown     ACTOR   
85399          Rutledge Wood         unknown     ACTOR   
85410       Zez?? di Camargo         unknown     ACTOR   
85411        Wanessa Camargo         unknown     ACTOR   
85429     Pranapong Khaisang             Tod     ACTOR   
85430  Apinya Sakuljaroensuk         Saiparn     ACTOR   
85431         Mawin Taweepol             Win     ACTOR   
85432          Nat Sakdatorn         unknown     ACTOR   
85433       Maneerat Kam-Uan              Ae     ACTOR   
85434      Rudklao Amratisha         unknown     ACTOR   
85435        Jaturong Mokjok         unknown     ACTOR   
85436       Pisamai Wilaisak         unknown     ACTOR   
85437    Sarawut Wichiensarn         unknown  DIRECTOR   

                                              title  type  release_year  \
77653                                     Let's Eat  SHOW          2021   
77654                                     Let's Eat  SHOW          2021   
77655                                     Let's Eat  SHOW          2021   
77656                                     Let's Eat  SHOW          2021   
77657                                     Let's Eat  SHOW          2021   
77658                                     Let's Eat  SHOW          2021   
77659                                     Let's Eat  SHOW          2021   
83911                     Dad Stop Embarrassing Me!  SHOW          2021   
84679                              Kayko and Kokosh  SHOW          2021   
84680                              Kayko and Kokosh  SHOW          2021   
84681                              Kayko and Kokosh  SHOW          2021   
84824                      Abla Fahita: Drama Queen  SHOW          2021   
84825                      Abla Fahita: Drama Queen  SHOW          2021   
84826          

The scores that are to be grouped should be rounded. For instance, titles with scores like 7.8, 8.1, and 8.3 will all be placed in the same bucket with a score of 8.

In [83]:
# rounding column with scores
df['imdb_score'] = df['imdb_score'].round(0)

#checking the outcome with tail()
print(df['imdb_score'].head(100))

1664    7.0
2076    9.0
2077    9.0
2078    9.0
2079    9.0
       ... 
4620    8.0
4621    8.0
4622    8.0
4623    8.0
4624    8.0
Name: imdb_score, Length: 100, dtype: float64


It is now time to identify outliers based on the number of votes.

In [84]:
# Use groupby() for scores and count all unique values in each group, print the result
unique_vote_counts = df.groupby('imdb_score')['imdb_score'].count()
#df.value_counts('imdb_score')
print(unique_vote_counts)

imdb_score
2.0       24
3.0       27
4.0      180
5.0      592
6.0     2494
7.0     4706
8.0     4842
9.0      557
10.0       8
Name: imdb_score, dtype: int64


<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Well done!
</div>

Based on the aggregation performed, it is evident that scores 2 (24 voted shows), 3 (27 voted shows), and 10 (only 8 voted shows) are outliers. There isn't enough data for these scores for the average number of votes to be meaningful.

To obtain the mean numbers of votes for the selected scores (we identified a range of 4-9 as acceptable), use conditional filtering and grouping.

In [85]:
# filter dataframe using two conditions (scores to be in the range 4-9)
filtered_df = df[(df['imdb_score'] >= 4) & (df['imdb_score'] <= 9)]
# group scores and corresponding average number of votes, reset index and print the result
avg_votes = filtered_df.groupby('imdb_score')['imdb_votes'].mean().reset_index()

print("Avg votes: ")
print(avg_votes)

Avg votes: 
   imdb_score     imdb_votes
0         4.0    5277.583333
1         5.0    3143.942568
2         6.0    3481.717322
3         7.0    8727.068211
4         8.0   30299.460967
5         9.0  126904.109515


Now for the final step! Round the column with the averages, rename both columns, and print the dataframe in descending order.

In [86]:
# round column with averages
print(round(avg_votes, 2))
print('\n')

# rename columns
df = df.rename(columns={
    'imdb_score':'score',
    'imdb_votes': 'votes'
    })

# print dataframe in descending order
display(df.sort_values(by='score', ascending=False))

   imdb_score  imdb_votes
0         4.0     5277.58
1         5.0     3143.94
2         6.0     3481.72
3         7.0     8727.07
4         8.0    30299.46
5         9.0   126904.11




name         character   role              title  type  \
4456   Bryan Cranston      Walter White  ACTOR       Breaking Bad  SHOW   
4457       Aaron Paul     Jesse Pinkman  ACTOR       Breaking Bad  SHOW   
4458        Anna Gunn      Skyler White  ACTOR       Breaking Bad  SHOW   
4459      Dean Norris     Hank Schrader  ACTOR       Breaking Bad  SHOW   
4460   Jonathan Banks  Mike Ehrmantraut  ACTOR       Breaking Bad  SHOW   
...               ...               ...    ...                ...   ...   
72745   Nikita Dragun     Nikita Dragun  ACTOR         Hype House  SHOW   
72744    Chase Hudson         Lil Huddy  ACTOR         Hype House  SHOW   
72740   Thomas Petrou     Thomas Petrou  ACTOR         Hype House  SHOW   
72741          Larray            Larray  ACTOR         Hype House  SHOW   
42330      Phoebe Lin       Wu Yi-Chuan  ACTOR  A House of Blocks  SHOW   

       release_year                          genres  score      votes  \
4456           2008  ['drama', 'crime', 'thriller']   10.0  1775990.0   
4457           2008  ['drama', 'crime', 'thriller']   10.0  1775990.0   
4458           2008  ['drama', 'crime', 'thriller']   10.0  1775990.0   
4459           2008  ['drama', 'crime', 'thriller']   10.0  1775990.0   
4460           2008  ['drama', 'crime', 'thriller']   10.0  1775990.0   
...             ...                             ...    ...        ...   
72745          2022                     ['reality']    2.0     2915.0   
72744          2022                     ['reality']    2.0     2915.0   
72740          2022                     ['reality']    2.0     2915.0   
72741          2022                     ['reality']    2.0     2915.0   
42330          2017             ['drama', 'horror']    2.0      137.0   

      imdb_rating  
4456         high  
4457         high  
4458         high  
4459         high  
4460         high  
...           ...  
72745         low  
72744         low  
72740         low  
72741         low  
42330         low  

[13430 rows x 10 columns]

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

OK, but we need only avg_votes dataframe here =)
</div>

The assumption macthes the analysis: the shows with the top 3 scores have the most amounts of votes.

## Conclusion <a id='hypotheses'></a>

The research done confirms that highly-rated shows released during the "Golden Age" of television also have the most votes. While shows with score 4 have more votes than ones with scores 5 and 6, the top three (scores 7-9) have the largest number. The data studied represents around 94% of the original set, so we can be confident in our findings.

<div class="alert alert-block alert-success">
<b>Overall reviewer's comment</b> <a class="tocSkip"></a>
    
Thank you for sending your project. You've done a really good job on it!

While there's room for improvement, on the whole, your project is impressive good. I like code style - very high level! Thank you for additional analysis. You did a great job! 
    
Keep up the good work, and good luck on the next sprint!
</div>